src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,29 @@
+package exercise2
+
+import java.util.EmptyStackException
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+class Stack[T] (elem: Option[T] = None, tail: Option[Stack[T]] = None) extends StackLike[T] {
```

Comment:
A more neat way to implement the stack is to split it into an `EmptyStack` and `Stack`. 
so 
```scala
case class Stack(elem: T, tail:StackLike[T]) extends StackLike[T]
```
and 
```scala
case class EmptyStack[T]() extends StackLike[T]
``` 
or even better
```scala
case object EmptyStack extends StackLike[Nothing]
```
but this has some complications with types, for which we need variance.
I go over this approach briefly in the upcoming lecture.